File: rankndcg.m
function [ndcg]=rankndcg(U,V,X,testset)
    para=U*V;
    ndcg=zeros(size(testset,1),1);
    for i=1:size(testset,1)
        testmat=testset{i};
        if (size(testmat,1)<200)
		continue;
	end
	w=para(:,testmat(1,1));
        rating=testmat(:,3)';
        rel=w'*X(:,testmat(:,2));
        [sortedrel,indexrel]=sort(rel,'descend');
        newrating=rating(indexrel);
        [sortedrat,indexrat]=sort(rating,'descend');
        
        sumrel=0;
        sumirel=0;
        for j=1:200
            sumrel=sumrel+(2.^(newrating(j))-1)/log2(j+1);
            sumirel=sumirel+(2.^(sortedrat(j))-1)/log2(j+1);
        end

        ndcg(i)=sumrel/sumirel;
    end
end

File: julia_version/rankndcg.jl
function rankndcg(U,V,W,m,X,testset,k,flag)
	para=U*V
	if flag!=0
		para=W
	end
        testparidx=parindex(testset)
	ndcg=zeros(m,1)
	dcg=0.0
	tdcg=0.0
	for i=1:m
		if flag==1
			w=W
		elseif flag==2
			w=W[:,i]
		else
			w=para[:,i]
		end
		rating=(X'*w)
		index=testparidx[i]+1:testparidx[i+1]
		realrating=testset[index,3]
		movieindex=testset[index,2]
		rating_use=rating[round(Int64,movieindex)]
		sortindex=sortperm(rating_use,rev=true)
		score=realrating[sortindex]
		sort!(realrating,rev=true)
		for j=1:k
			if testparidx[i+1]-testparidx[i]<j
				break;
			end
			dcg+=(2^score[j]-1)/log2(j+1)
			tdcg+=(2^realrating[j]-1)/log2(j+1)
		end
	end
	return dcg/tdcg
end
